---
title: "R Notebook"
output: html_notebook
---

```{r}

library(Seurat)
library(SeuratDisk)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(clustree)
library(reshape2)
library(tidyr)
library(tidyverse)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/xgboost_train.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/plottingFxns.R")

obj.opossum.glutamatergic <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_orig_glutamatergic.rds")

```


```{r}

obj.opossum.glutamatergic <- ClusterSCT(obj.opossum.glutamatergic, c(0.1))
PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.1")

```


```{r}

canon.markers <- list(
                    class = c("Slc17a6", "Slc17a7", "Sv2b"),
                    L23 = c("Cux2", "Ccbe1", "Mdga1", "Stard8"),
                    L4 = c("Whrn", "Rorb"),
                    L5IT = c("Bcl11b"),
                    L5NP = c("Nxph1", "Tshz2", "Trhr", "Slc17a8"),
                    L5PT = c("Bcl6", "Erg", "Reln"),
                    L6CT = c("Foxp2", "Syt6"),
                    L6IT = c("Zfp804b", "Cdh9"),
                    L6b = c("Ctgf", "Inpp4b", "Svil")
                      )

# look for canonical markers
DefaultAssay(obj.opossum.glutamatergic) <- "RNA"
Idents(obj.opossum.glutamatergic) <- "SCT_snn_res.0.1"
PlotFeatures(obj.opossum.glutamatergic, canon.markers)
DotPlot(obj.opossum.glutamatergic, features = canon.markers, cols = c("lightgrey", "red"), scale = FALSE) + 
                         theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
                         theme(panel.background = element_rect(fill = "white", color = NA), plot.background = element_rect(fill = "white", color = NA)) + NoLegend()

```


```{r}

obj.opossum.glutamatergic.prefilt <- obj.opossum.glutamatergic
obj.opossum.glutamatergic <- subset(obj.opossum.glutamatergic, idents = c(6), invert = TRUE)

```


```{r}

obj.opossum.glutamatergic <- ClusterSCT(obj.opossum.glutamatergic, c(0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.5, 1, 1.5))
PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.1")

canon.markers <- list(
                    class = c("Slc17a6", "Slc17a7", "Sv2b"),
                    L23 = c("Cux2", "Ccbe1", "Mdga1", "Stard8"),
                    L4 = c("Whrn", "Rorb"),
                    L5IT = c("Bcl11b"),
                    L5NP = c("Nxph1", "Tshz2", "Trhr", "Slc17a8"),
                    L5PT = c("Bcl6", "Erg", "Reln"),
                    L6CT = c("Foxp2", "Syt6"),
                    L6IT = c("Zfp804b", "Cdh9"),
                    L6b = c("Ctgf", "Inpp4b", "Svil")
                      )

# look for canonical markers
DefaultAssay(obj.opossum.glutamatergic) <- "RNA"
Idents(obj.opossum.glutamatergic) <- "SCT_snn_res.0.1"
PlotFeatures(obj.opossum.glutamatergic, canon.markers)
DotPlot(obj.opossum.glutamatergic, features = canon.markers, cols = c("lightgrey", "red"), scale = FALSE) + 
                         theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
                         theme(panel.background = element_rect(fill = "white", color = NA), plot.background = element_rect(fill = "white", color = NA)) + NoLegend()

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.1")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.15")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.2")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.25")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.3")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.35")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.0.5")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.1")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.1.5")

```


```{r, fig.height=10, fig.width=10}

clustree(obj.opossum.glutamatergic, prefix = "SCT_snn_res.")

```


```{r}

p <- DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = "SCT_snn_res.0.2", label = TRUE, raster = FALSE) + NoLegend() + xlim(-13, 15) + ylim(-14, 14) + coord_equal()
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_UMAP_res.0.2.svg", plot = p, width = 5, height = 5, dpi = 300)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_UMAP_res.0.2.png", plot = p, width = 5, height = 5, dpi = 300)
p <- DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = "SCT_snn_res.0.3", label = TRUE, raster = FALSE) + NoLegend() + xlim(-13, 15) + ylim(-14, 14) + coord_equal()
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_UMAP_res.0.3.svg", plot = p, width = 5, height = 5, dpi = 300)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_UMAP_res.0.3.png", plot = p, width = 5, height = 5, dpi = 300)
p <- DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = "SCT_snn_res.0.4", label = TRUE, raster = FALSE) + NoLegend() + xlim(-13, 15) + ylim(-14, 14) + coord_equal()
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_UMAP_res.0.4.svg", plot = p, width = 5, height = 5, dpi = 300)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_UMAP_res.0.4.png", plot = p, width = 5, height = 5, dpi = 300)

```


```{r, fig.width = 8, fig.height = 8}

source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")

obj.opossum.glutamatergic$subclass.0.2 <- obj.opossum.glutamatergic$SCT_snn_res.0.2
p <- SaveSubclassConfusionMatrices(obj.opossum.glutamatergic, "subclass.0.2",
                              c("1", "2", "5", "4", "7", "6", "3"),
                              "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/", 
                              return = TRUE, colormap_upper_limit = 10)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_Classification_res.0.2.svg", plot = p, width = 5, height = 5, dpi = 300)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_Classification_res.0.2.png", plot = p, width = 5, height = 5, dpi = 300)

obj.opossum.glutamatergic$subclass.0.3 <- obj.opossum.glutamatergic$SCT_snn_res.0.3
p <- SaveSubclassConfusionMatrices(obj.opossum.glutamatergic, "subclass.0.3",
                              c("1", "2", "5", "6", "4", "9", "8", "7", "3", "10"),
                              "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/", 
                              return = TRUE, colormap_upper_limit = 10)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_Classification_res.0.3.svg", plot = p, width = 6, height = 6, dpi = 300)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_Classification_res.0.3.png", plot = p, width = 6, height = 6, dpi = 300)

obj.opossum.glutamatergic$subclass.0.4 <- obj.opossum.glutamatergic$SCT_snn_res.0.4
p <- SaveSubclassConfusionMatrices(obj.opossum.glutamatergic, "subclass.0.4",
                              c("2", "3", "5", "4", "9", "6", "8", "10", "12", "11", "7", "1", "13"),
                              "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/", 
                              return = TRUE, colormap_upper_limit = 10)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_Classification_res.0.4.svg", plot = p, width = 8, height = 8, dpi = 300)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Glutamatergic_Classification_res.0.4.png", plot = p, width = 8, height = 8, dpi = 300)

# obj.opossum.glutamatergic$subclass.0.2.merge <- obj.opossum.glutamatergic$subclass.0.2
# obj.opossum.glutamatergic$subclass.0.2.merge[obj.opossum.glutamatergic$subclass.0.2.merge == 2] <- 1
# 
# SaveSubclassConfusionMatrices(obj.opossum.glutamatergic, "subclass.0.2.merge",
#                               c("1", "3", "4", "5", "6", "7"),
#                               "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/")

```


```{r}

source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_integration_functions.R")
library(plyr)

objs <- list(Opossum = obj.opossum.glutamatergic, Mouse = obj.mouse.glutamatergic)
objs$Opossum$subclass <- objs$Opossum$subclass.0.2

objs.m <- MapObjects(objs[["Opossum"]], objs[["Mouse"]], c("subclass"), assay = "SCT")

p <- PlotMappedLabelsHeatmap(objs.m[[1]], "subclass", c("L2/3", "L4", "L5IT", "L6IT", "L5NP", "L5PT", "L6CT", "L6b"), normalize = "row", 
                        ident.order = c("1", "2", "5", "4", "7", "6", "3"))

```


```{r}

classes <- c("Glutamatergic")
iters <- c()
objs.opossum = list()
objs.mouse = list()
mapping.data = list("cell.id" = c(), "subclass" = c(), "predicted.subclass" = c(), "predicted.subclass.score" = c())
mapping.classes = list("Glutamatergic" = mapping.data, "GABAergic" = mapping.data, "Nonneuronal" = mapping.data)
mapping.species <- list("Opossum" = mapping.classes, "Mouse" = mapping.classes)
objs.opossum$Glutamatergic <- obj.opossum.glutamatergic
objs.mouse$Glutamatergic <- obj.mouse.glutamatergic

objs.opossum$Glutamatergic$subclass <- objs.opossum$Glutamatergic$subclass.0.4 #### set subclass

iteration.list = list("Opossum" = list(), "Mouse" = list())
for (cl in classes) {
  iteration.list$Opossum[[cl]] <- SubsampleObjectMultipleIterations(objs.opossum[[cl]], "subclass", 100, 100)
  iteration.list$Mouse[[cl]] <- SubsampleObjectMultipleIterations(objs.mouse[[cl]], "subclass", 100, 100)
}

for (cl in classes) {
  for (it in 1:10) {

    obj.opossum <- objs.opossum[[cl]]
    obj.opossum$species <- "Opossum"
    obj.mouse <- objs.mouse[[cl]]
    obj.mouse$species <- "Mouse"
    # obj.opossum <- SubsampleObject(obj.opossum, "subclass", 200)
    obj.opossum <- subset(obj.opossum, cells = as.character(iteration.list$Opossum[[cl]][[it]]))
    # obj.mouse <- SubsampleObject(obj.mouse, "subclass", 200)
    obj.mouse <- subset(obj.mouse, cells = as.character(iteration.list$Mouse[[cl]][[it]]))
    objs <- list(obj.opossum, obj.mouse)
    objs.m <- MapObjects(objs[[1]], objs[[2]], c("subclass"), assay = "SCT")
    mapping.species$Opossum[[cl]][["cell.id"]] <- c(mapping.species$Opossum[[cl]][["cell.id"]], as.character(colnames(objs.m[[1]])))
    mapping.species$Opossum[[cl]][["subclass"]] <- c(mapping.species$Opossum[[cl]][["subclass"]], as.character(objs.m[[1]]$subclass))
    mapping.species$Opossum[[cl]][["predicted.subclass"]] <- c(mapping.species$Opossum[[cl]][["predicted.subclass"]], as.character(objs.m[[1]]$predicted.subclass))
    mapping.species$Opossum[[cl]][["predicted.subclass.score"]] <- c(mapping.species$Opossum[[cl]][["predicted.subclass.score"]], as.numeric(objs.m[[1]]$predicted.subclass.score))
    mapping.species$Mouse[[cl]][["cell.id"]] <- c(mapping.species$Mouse[[cl]][["cell.id"]], as.character(colnames(objs.m[[2]])))
    mapping.species$Mouse[[cl]][["subclass"]] <- c(mapping.species$Mouse[[cl]][["subclass"]], as.character(objs.m[[2]]$subclass))
    mapping.species$Mouse[[cl]][["predicted.subclass"]] <- c(mapping.species$Mouse[[cl]][["predicted.subclass"]], as.character(objs.m[[2]]$predicted.subclass))
    mapping.species$Mouse[[cl]][["predicted.subclass.score"]] <- c(mapping.species$Mouse[[cl]][["predicted.subclass.score"]], as.numeric(objs.m[[2]]$predicted.subclass.score))
  
  }
}

p <- PlotSubsampledMappedLabelsHeatmap(mapping.species$Opossum$Glutamatergic$subclass, mapping.species$Opossum$Glutamatergic$predicted.subclass, 
                                  c("L2/3", "L4", "L5IT", "L6IT", "L5NP", "L5PT", "L6CT", "L6b"), normalize = "row", 
                                  ident.order = c("2", "3", "5", "4", "9", "6", "8", "10", "12", "11", "7", "1", "13", "L2/3", "L4", "L5IT", "L6IT", "L5NP", "L5PT", "L6CT", "L6b"))
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Mouse_Glutamatergic_Classification_res.0.4.svg", plot = p, width = 5, height = 7, dpi = 300)
ggsave("E:/Opossum_Paper/Figure S2/Opossum_Mouse_Glutamatergic_Classification_res.0.4.png", plot = p, width = 5, height = 7, dpi = 300)

```


```{r}

obj.mouse.glutamatergic <- readRDS("E:/Transcriptomics_V1/Mouse/seurat/mouse_v1_P38_glutamatergic.rds")

Idents(obj.opossum.glutamatergic) <- "SCT_snn_res.0.2"
obj.opossum.IT <- subset(obj.opossum.glutamatergic, idents = c(1, 2, 4, 5, 7))

# Step 1: Obtain DE genes for opossum
markers_opossum <- FindAllMarkers(obj.opossum.IT)

# Step 2: Identify shared and unique genes
shared_genes <- intersect(rownames(obj.opossum.IT), rownames(obj.mouse.glutamatergic))
unique_opossum_genes <- setdiff(rownames(obj.opossum.IT), rownames(obj.mouse.glutamatergic))

# Step 3: Categorize DE genes as shared or unique
markers_opossum <- markers_opossum %>%
  mutate(
    gene_type = case_when(
      gene %in% shared_genes ~ "Shared",
      gene %in% unique_opossum_genes ~ "Unique to Opossum",
      TRUE ~ "Other" # This should not happen, but just for safety
    )
  )

# Summarize DE gene counts by cluster and gene type
de_gene_summary <- markers_opossum %>%
  group_by(cluster, gene_type) %>%
  summarise(count = n())

# Step 4: Create the stacked barplot
ggplot(de_gene_summary, aes(x = cluster, y = count, fill = gene_type)) +
  geom_bar(stat = "identity") +
  labs(x = "Opossum Cluster", y = "Number of DE Genes", fill = "Gene Type") +
  theme_minimal()

```


```{r}

# identify subclasses at each clustering resolution
# do differential expression for clusters within each subclass
# generate plots of top genes for each

subclass.idx.opossum <- list()

subclass.idx.opossum$SCT_snn_res.0.2$IT_A <- c("1", "2")
subclass.idx.opossum$SCT_snn_res.0.3$IT_A <- c("1", "2", "5")
subclass.idx.opossum$SCT_snn_res.1$IT_A <- c("3", "4", "7", "11", "12", "13", "14", "15")

subclass.idx.opossum$SCT_snn_res.0.2$IT_B <- c("4")
subclass.idx.opossum$SCT_snn_res.0.3$IT_B <- c("4")
subclass.idx.opossum$SCT_snn_res.1$IT_B <- c("6", "10", "19")

subclass.idx.opossum$SCT_snn_res.0.2$IT_C <- c("5")
subclass.idx.opossum$SCT_snn_res.0.3$IT_C <- c("6")
subclass.idx.opossum$SCT_snn_res.1$IT_C <- c("2", "9")

subclass.idx.opossum$SCT_snn_res.0.2$IT_D <- c("7")
subclass.idx.opossum$SCT_snn_res.0.3$IT_D <- c("9")
subclass.idx.opossum$SCT_snn_res.1$IT_D <- c("17")

subclass.idx.opossum$SCT_snn_res.0.2$L5PT <- c("6")
subclass.idx.opossum$SCT_snn_res.0.3$L5PT <- c("7")
subclass.idx.opossum$SCT_snn_res.1$L5PT <- c("5")

subclass.idx.opossum$SCT_snn_res.0.3$L5NP <- c("8")
subclass.idx.opossum$SCT_snn_res.1$L5NP <- c("16")

subclass.idx.opossum$SCT_snn_res.0.2$L6CT <- c("3")
subclass.idx.opossum$SCT_snn_res.0.3$L6CT <- c("3")
subclass.idx.opossum$SCT_snn_res.1$L6CT <- c("1", "8")

subclass.idx.opossum$SCT_snn_res.0.3$L6b <- c("10")
subclass.idx.opossum$SCT_snn_res.1$L6b <- c("20")

obj.opossum.glutamatergic <- SubclassByIdent(obj.opossum.glutamatergic, subclass.idx.opossum)

```


```{r}

subclass.labels <- c("IT_A", "IT_B", "IT_C", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.0.2", "SCT_snn_res.0.3", "SCT_snn_res.1")

markers.opossum.glutamatergic <- IdentMarkerDict(obj.opossum.glutamatergic, subclass.labels, ident.labels,
                                        "E:/Transcriptomics_V1/Opossum/seurat/marker_genes/Glutamatergic/Original/markerdict_clusters.rds")
SaveDotPlots(obj.opossum.glutamatergic, markers.opossum.glutamatergic,
             subclass.labels, ident.labels,
             "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/", "ENSMODG")
SaveFeaturePlots(obj.opossum.glutamatergic, markers.opossum.glutamatergic,
                 subclass.labels, ident.labels,
                 "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/")

```


```{r}

subclass.labels <- c("IT_A", "IT_B", "IT_C", "IT_D", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.0.2", "SCT_snn_res.0.5", "SCT_snn_res.1.5")

for (sbcl in subclass.labels) {
  for (id in ident.labels) {
    p <- PlotIdentGeneCounts(markers.opossum.glutamatergic, sbcl, id)
    p <- p + theme(aspect.ratio = 1) # Make plot square
    print(p)
  }
}

```


```{r}

DefaultAssay(obj.opossum.glutamatergic) <- "SCT"
ident = "SCT_snn_res.2"
gene <- "Pvalb"
DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = ident, label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
FeaturePlot(obj.opossum.glutamatergic, gene, raster = FALSE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
VlnPlot(obj.opossum.glutamatergic, gene, group.by = ident)

```


```{r}

# whole-transcriptome cross-validated classification
subclass.labels <- c("IT_A", "IT_B", "IT_C", "IT_D", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.0.2", "SCT_snn_res.0.5", "SCT_snn_res.1.5")

SaveIdentConfusionMatrices(obj.opossum.glutamatergic, subclass.labels, ident.labels, 
                      "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/")

```


```{r}

# label subclasses and types
obj.opossum.glutamatergic$subclass <- NA
obj.opossum.glutamatergic$type <- NA

subclass.resolutions <- rev(list(IT_A = 0.5, IT_B = 0.5, IT_C = 0.5, IT_D = 0.5, L5PT = 1.5, L5NP = 0.5, L6CT = 1.5, L6b = 0.5))
obj.opossum.glutamatergic <- LabelCells(obj.opossum.glutamatergic, subclass.resolutions)

```


```{r}

DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

```


```{r}

# Plot relative subclass proportions
Idents(obj.opossum.glutamatergic) <- "subclass"
levels(obj.opossum.glutamatergic) <- c("IT_A", "IT_B", "IT_C", "IT_D", "L5PT", "L5NP", "L6CT", "L6b")

IdentBySample(obj.opossum.glutamatergic, y_limits = c(0, 0.60))

```


```{r}

subclass.labels <- c("IT_A", "IT_B", "IT_C", "L5PT", "L6CT")
ident.labels <- c("type")

subclass.markers.opossum.glutamatergic <- SubclassMarkerDict(obj.opossum.glutamatergic, "subclass",
                                                       "E:/Transcriptomics_V1/Opossum/seurat/marker_genes/Glutamatergic/Original/markerdict_subclass.rds")
markers.opossum.glutamatergic <- IdentMarkerDict(obj.opossum.glutamatergic, subclass.labels, ident.labels,
                                        "E:/Transcriptomics_V1/Opossum/seurat/marker_genes/Glutamatergic/Original/markerdict_types.rds")
SaveDotPlots(obj.opossum.glutamatergic, markers.opossum.glutamatergic,
             subclass.labels, ident.labels,
             "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/", "ENSMODG")
SaveFeaturePlots(obj.opossum.glutamatergic, markers.opossum.glutamatergic,
                 subclass.labels, ident.labels,
                 "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/")

PlotSubclassGeneCounts(subclass.markers.opossum.glutamatergic, "subclass", 
                       c("IT_A", "IT_B", "IT_C", "IT_D", "L5PT", "L5NP", "L6CT", "L6b")) + theme(aspect.ratio = 1)
for (sbcl in subclass.labels) {
  for (id in ident.labels) {
    p <- PlotIdentGeneCounts(markers.opossum.glutamatergic, sbcl, id)
    p <- p + theme(aspect.ratio = 1) # Make plot square
    print(p)
  }
}

SaveSubclassConfusionMatrices(obj.opossum.glutamatergic, "subclass",
                              c("IT_A", "IT_B", "IT_C", "IT_D", "L5PT", "L5NP", "L6CT", "L6b"),
                              "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/")
SaveIdentConfusionMatrices(obj.opossum.glutamatergic, subclass.labels, ident.labels,
                           "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/Original/")

```


```{r}

saveRDS(obj.opossum.glutamatergic, "E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_orig_glutamatergic_processed.rds")

```

