---
title: "Integrate GABAergic Mouse Opossum"
output: html_notebook
---

```{r}

library(Seurat)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(cowplot)
library(dplyr)
library(SeuratDisk)
library(lisi)

obj.mouse <- LoadH5Seurat("E:/Mouse_M1/seurat/mouse_m1_gabaergic.h5seurat")
# obj.mouse[["dataset"]] <- "Mouse_M1"
# obj.mouse[["subclass"]] <- paste0(obj.mouse$subclass_label, " Mouse")

obj.opossum <- LoadH5Seurat("E:/Opossum_M1/seurat/opossum_m1_gabaergic_raw.h5seurat")
obj.opossum[["dataset"]] <- "Opossum_M1"
# obj.opossum[["subclass"]] <- paste0(obj.opossum$subclass, " Opossum")

```


```{r}

common.features <- intersect(rownames(obj.mouse), rownames(obj.opossum))
obj.combined <- merge(obj.mouse[common.features,], y = obj.opossum[common.features,])
obj.list <- SplitObject(obj.combined, split.by = "dataset")

```


```{r}

obj.list[["Mouse_M1"]] <- SCTransform(obj.list[["Mouse_M1"]], vst.flavor = "v2", verbose = FALSE) %>%
               RunPCA(npcs = 30, verbose = FALSE) %>%
               RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
               FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
               FindClusters(resolution = 1, algorithm = 4, method = "igraph")

DimPlot(obj.list[["Mouse_M1"]], label = TRUE) + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
DimPlot(obj.list[["Mouse_M1"]], label = TRUE, group.by = "subclass") + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
DimPlot(obj.list[["Mouse_M1"]], label = TRUE, group.by = "type") + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()

obj.list[["Opossum_M1"]] <- SCTransform(obj.list[["Opossum_M1"]], vst.flavor = "v2", verbose = FALSE) %>%
               RunPCA(npcs = 30, verbose = FALSE) %>%
               RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
               FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
               FindClusters(resolution = 1, algorithm = 4, method = "igraph")

DimPlot(obj.list[["Opossum_M1"]], label = TRUE) + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
# DimPlot(obj.list[["Opossum_M1"]], label = TRUE, group.by = "subclass") + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
# DimPlot(obj.list[["Opossum_M1"]], label = TRUE, group.by = "type") + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()

```


```{r}

features <- SelectIntegrationFeatures(object.list = obj.list, nfeatures = 5000)
obj.list <- PrepSCTIntegration(object.list = obj.list, anchor.features = features)
anchors <- FindIntegrationAnchors(object.list = obj.list, normalization.method = "SCT", anchor.features = features)
obj.combined.sct <- IntegrateData(anchorset = anchors, normalization.method = "SCT", features.to.integrate = common.features)

```


```{r}

obj.combined.sct <- RunPCA(obj.combined.sct, verbose = FALSE)
obj.combined.sct <- RunUMAP(obj.combined.sct, reduction = "pca", dims = 1:30, verbose = FALSE)
obj.combined.sct <- FindNeighbors(obj.combined.sct, reduction = "pca", dims = 1:30)
obj.combined.sct <- FindClusters(obj.combined.sct, resolution = 1, algorithm = 4, method = "igraph")

```


```{r}

DimPlot(obj.combined.sct, reduction = "umap", group.by = "dataset", raster = FALSE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.combined.sct, reduction = "umap", group.by = "subclass", split.by = "dataset", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
# DimPlot(obj.combined.sct, reduction = "umap", group.by = "type", split.by = "dataset", label = TRUE, raster = FALSE, label.size = 1) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
# DimPlot(obj.combined.sct, reduction = "umap", group.by = "RNA_snn_res.1", split.by = "dataset", label = TRUE, raster = FALSE, label.size = 2) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.combined.sct, reduction = "umap", group.by = "integrated_snn_res.1", split.by = "dataset", label = TRUE, raster = FALSE, label.size = 2) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

```


```{r}

obj.combined.sct.mrk <- PrepSCTFindMarkers(obj.combined.sct, assay = "SCT", verbose = TRUE)

```


```{r}

FeaturePlot(obj.combined.sct.mrk, reduction = "umap", features = c("Sncg"), split.by = "dataset", raster = FALSE, min.cutoff = 0, max.cutoff = 5) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

```


```{r}

FeaturePlot(obj.list[["Opossum_M1"]], reduction = "umap", features = c("Sncg"), raster = FALSE, min.cutoff = 0, max.cutoff = 5) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

```


```{r}

pvalb.mouse.palette <- colorRampPalette(c("#ffffdd", "#468fba"))(8)[3:8]
sst.mouse.palette <- colorRampPalette(c("#ffffdd", "#ba1851"))(9)[2:9]
vip.mouse.palette <- colorRampPalette(c("#ffffdd", "#a85d96"))(7)[2:7]
lamp5.mouse.palette <- colorRampPalette(c("#ffffdd", "#a2bc8c"))(5)[2:5]
sncg.mouse.palette <- colorRampPalette(c("#ffffdd", "#f6cb7c"))(6)[c(2, 4, 6)]
meis2.mouse.palette <- "darkgray"

mouse.levels <- c("Pvalb Il1rapl2", "Pvalb Reln", "Pvalb Vipr2", "Pvalb Th", "Pvalb Htr2c", "Pvalb Sema3e", 
                  "Sst Crhr2", "Sst Nr2f2", "Sst Chodl", "Sst Necab1", "Sst Calb2", "Sst Etv1", "Sst Hpse", "Sst Crh", 
                  "Vip Npy", "Vip Ptger3", "Vip Mybpc1", "Vip Chat", "Vip Col14a1", "Vip Reln", 
                  "Lamp5 Lhx6", "Lamp5 Ndnf", "Lamp5 Cdh4", "Lamp5 Htr1f", 
                  "Sncg Slc17a8", "Sncg Cpne5", "Sncg Vip", 
                  "Meis2")

pvalb.opossum.palette <- colorRampPalette(c("#ffffdd", "#468fba"))(8)[2:8]
sst.opossum.palette <- colorRampPalette(c("#ffffdd", "#ba1851"))(8)[c(3, 5, 6, 7, 8)]
vip.opossum.palette <- colorRampPalette(c("#ffffdd", "#a85d96"))(6)[c(2, 4, 6)]
lamp5.opossum.palette <- colorRampPalette(c("#ffffdd", "#a2bc8c"))(6)[c(6)]

opossum.levels <- c("Pvalb Plod2", "Pvalb Cemip", "Pvalb Trpc6", "Pvalb Ltbp1", "Pvalb Crhbp", "Pvalb Col25a1", "Pvalb Ptpn7", 
                    "Sst Ptpn7", "Sst Plpp4", "Sst Vcan", "Sst Htr2a", "Sst Col8a1", 
                    "Vip Mid1", "Vip Abi3bp", "Vip Rxfp1", 
                    "Lamp5 Kit")

Idents(obj.combined.sct) <- "type"
levels(obj.combined.sct) <- c(mouse.levels, opossum.levels)
colors <- c(pvalb.mouse.palette, sst.mouse.palette, vip.mouse.palette, lamp5.mouse.palette, sncg.mouse.palette, meis2.mouse.palette, 
            pvalb.opossum.palette, sst.opossum.palette, vip.opossum.palette, lamp5.opossum.palette)

dim_plot <- DimPlot(obj.combined.sct, reduction = "umap", split.by = "dataset", cols = colors, label = TRUE, raster = FALSE, label.size = 2) + NoLegend() + xlim(-15, 16) + ylim(-16, 16) + coord_equal()

ggsave("G:/Shared drives/Opossum transcriptomics/figures/Opossum_Mouse_Types_UMAP_GABAergic.png", dim_plot, width=7, height=4, dpi=500)

```


```{r}

Idents(obj.combined.sct) <- "integrated_snn_res.1"
obj.combined.sct <- PrepSCTFindMarkers(obj.combined.sct)
markers <- FindConservedMarkers(obj.combined.sct, assay = "SCT", ident.1 = "22", grouping.var = "dataset", verbose = FALSE)

```


```{r}

Idents(obj.combined.sct) <- "subclass"
obj.combined.sct.sst <- subset(obj.combined.sct, idents = "Sst")
obj.combined.sct.sst <- PrepSCTFindMarkers(obj.combined.sct.sst)
Idents(obj.combined.sct.sst) <- "dataset"
sst.markers <- FindMarkers(obj.combined.sct.sst, assay = "SCT", ident.1 = "Opossum_M1", ident.2 = "Mouse_M1", verbose = FALSE)

```


```{r}

library(clustifyr)
expr_mat <- obj.combined.sct[["integrated"]]@data
ref_mat <- average_clusters(mat = expr_mat, metadata = obj.combined.sct[[]], cluster_col = "subclass", if_log = TRUE)

res <- clustify(
  input = expr_mat, # matrix of normalized scRNA-seq counts (or SCE/Seurat object)
  metadata = obj.combined.sct[[]], # meta.data table containing cell clusters
  cluster_col = "subclass", # name of column in meta.data containing cell clusters
  ref_mat = ref_mat, # matrix of RNA-seq expression data for each cell type
  query_genes = rownames(obj.combined.sct) # list of highly varible genes identified with Seurat
                )

```


```{r}

library(clustifyr)
expr_mat <- t(Embeddings(obj.combined.sct, reduction = "pca"))
ref_mat <- average_clusters(mat = expr_mat, metadata = obj.combined.sct[[]], cluster_col = "subclass", if_log = TRUE)

res <- clustify(
  input = expr_mat, # matrix of normalized scRNA-seq counts (or SCE/Seurat object)
  metadata = obj.combined.sct[[]], # meta.data table containing cell clusters
  cluster_col = "subclass", # name of column in meta.data containing cell clusters
  ref_mat = ref_mat, # matrix of RNA-seq expression data for each cell type
  query_genes = rownames(expr_mat) # list of highly varible genes identified with Seurat
                )

```


```{r, fig.height=4.5, fig.width=5}

order <- c("Pvalb Mouse", "Pvalb Opossum", 
           "Sst Mouse", "Sst Opossum", 
           "Sst Chodl Mouse", "Sst Chodl Opossum", 
           "Vip Mouse", "Vip Opossum", 
           "Lamp5 Mouse", "Lamp5 Opossum", 
           "Sncg Mouse", 
           "Meis2 Mouse")

# es.order <- res[order, order]

plot_cor_heatmap(cor_mat = res)

```

