---
title: "Integrate GABAergic Mouse Opossum"
output: html_notebook
---

```{r}

library(Seurat)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(cowplot)
library(dplyr)
library(SeuratDisk)

obj.mouse <- LoadH5Seurat("E:/Mouse_M1/seurat/mouse_m1_gabaergic.h5seurat")
obj.mouse[["dataset"]] <- "Mouse_M1"
obj.mouse[["subclass"]] <- paste0(obj.mouse$subclass_label, " M1")

obj.opossum <- LoadH5Seurat("E:/mouse_v1_gabaergic.h5seurat")
obj.opossum[["dataset"]] <- "Opossum_M1"
obj.opossum[["subclass"]] <- paste0(obj.opossum$subclass, " V1")

obj.mouse <- obj.mouse[, sample(colnames(obj.mouse), size = ncol(obj.opossum), replace = FALSE)]

```


```{r}

common.features <- intersect(rownames(obj.mouse), rownames(obj.opossum))
obj.combined <- merge(obj.mouse[common.features,], y = obj.opossum[common.features,])
obj.list <- SplitObject(obj.combined, split.by = "dataset")

```


```{r}

obj.list[["Mouse_M1"]] <- SCTransform(obj.list[["Mouse_M1"]], vst.flavor = "v2", verbose = FALSE) %>%
               RunPCA(npcs = 30, verbose = FALSE) %>%
               RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
               FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
               FindClusters(resolution = 1, algorithm = 4, method = "igraph")

DimPlot(obj.list[["Mouse_M1"]], label = TRUE, repel = TRUE) + xlim(-15, 15) + ylim(-15, 15) + coord_equal()

obj.list[["Opossum_M1"]] <- SCTransform(obj.list[["Opossum_M1"]], vst.flavor = "v2", verbose = FALSE) %>%
               RunPCA(npcs = 30, verbose = FALSE) %>%
               RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
               FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
               FindClusters(resolution = 1, algorithm = 4, method = "igraph")

DimPlot(obj.list[["Opossum_M1"]], label = TRUE, repel = TRUE) + xlim(-15, 15) + ylim(-15, 15) + coord_equal()

```


```{r}

features <- SelectIntegrationFeatures(object.list = obj.list, nfeatures = 3000)
obj.list <- PrepSCTIntegration(object.list = obj.list, anchor.features = features)
anchors <- FindIntegrationAnchors(object.list = obj.list, normalization.method = "SCT", anchor.features = features)
obj.combined.sct <- IntegrateData(anchorset = anchors, normalization.method = "SCT")

```


```{r}

obj.combined.sct <- RunPCA(obj.combined.sct, verbose = FALSE)
obj.combined.sct <- RunUMAP(obj.combined.sct, reduction = "pca", dims = 1:30, verbose = FALSE)
obj.combined.sct <- FindNeighbors(obj.combined.sct, reduction = "pca", dims = 1:30)
obj.combined.sct <- FindClusters(obj.combined.sct, resolution = 1, algorithm = 4, method = "igraph")

```


```{r}

# DimPlot(obj.combined.sct, reduction = "umap", group.by = "integrated_snn_res.1", label = TRUE) + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
DimPlot(obj.combined.sct, reduction = "umap", group.by = "dataset") + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
# ggsave(file="G:/Shared drives/Opossum transcriptomics/figures/Mouse_Opossum_M1_GABAergic_Dataset.svg", plot=dataset_plot, width=6, height=4)
# DimPlot(obj.combined.sct, reduction = "umap", group.by = "cluster_label", label = TRUE) + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
# DimPlot(obj.combined.sct, reduction = "umap", split.by = "dataset", group.by = "RNA_snn_res.1.5", label = TRUE) + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
DimPlot(obj.combined.sct, reduction = "umap", split.by = "dataset", group.by = "subclass", label = TRUE) + NoLegend() + xlim(-15, 15) + ylim(-15, 15) + coord_equal()
# ggsave(file="G:/Shared drives/Opossum transcriptomics/figures/Mouse_Opossum_M1_GABAergic_Subclass.svg", plot=subclass_plot, width=8.5, height=4.5)

```


```{r}

Idents(obj.combined.sct) <- "subclass"
obj.combined.sct.vip <- subset(obj.combined.sct, idents = "Vip")
obj.combined.sct.vip <- PrepSCTFindMarkers(obj.combined.sct.vip)
Idents(obj.combined.sct.vip) <- "dataset"
vip.markers <- FindMarkers(obj.combined.sct.vip, assay = "SCT", ident.1 = "Opossum_M1", ident.2 = "Mouse_M1", verbose = FALSE)

```


```{r}

Idents(obj.combined.sct) <- "subclass"
obj.combined.sct.sst <- subset(obj.combined.sct, idents = "Sst")
obj.combined.sct.sst <- PrepSCTFindMarkers(obj.combined.sct.sst)
Idents(obj.combined.sct.sst) <- "dataset"
sst.markers <- FindMarkers(obj.combined.sct.sst, assay = "SCT", ident.1 = "Opossum_M1", ident.2 = "Mouse_M1", verbose = FALSE)

```


```{r}

library(clustifyr)
expr_mat <- obj.combined.sct[["integrated"]]@data
ref_mat <- average_clusters(mat = expr_mat, metadata = obj.combined.sct[[]], cluster_col = "subclass", if_log = TRUE)

res <- clustify(
  input = expr_mat, # matrix of normalized scRNA-seq counts (or SCE/Seurat object)
  metadata = obj.combined.sct[[]], # meta.data table containing cell clusters
  cluster_col = "subclass", # name of column in meta.data containing cell clusters
  ref_mat = ref_mat, # matrix of RNA-seq expression data for each cell type
  query_genes = rownames(obj.combined.sct) # list of highly varible genes identified with Seurat
                )

```


```{r}

library(clustifyr)
expr_mat <- t(Embeddings(obj.combined.sct, reduction = "pca"))
ref_mat <- average_clusters(mat = expr_mat, metadata = obj.combined.sct[[]], cluster_col = "subclass", if_log = TRUE)

res <- clustify(
  input = expr_mat, # matrix of normalized scRNA-seq counts (or SCE/Seurat object)
  metadata = obj.combined.sct[[]], # meta.data table containing cell clusters
  cluster_col = "subclass", # name of column in meta.data containing cell clusters
  ref_mat = ref_mat, # matrix of RNA-seq expression data for each cell type
  query_genes = rownames(expr_mat) # list of highly varible genes identified with Seurat
                )

```


```{r, fig.height=4.5, fig.width=5}

order <- c("Pvalb Mouse", "Pvalb Opossum", 
           "Sst Mouse", "Sst Opossum", 
           "Sst Chodl Mouse", "Sst Chodl Opossum", 
           "Vip Mouse", "Vip Opossum", 
           "Lamp5 Mouse", "Lamp5 Opossum", 
           "Sncg Mouse", 
           "Meis2 Mouse")

# res.order <- res[order, order]
res[12, 12] <- -1
plot_cor_heatmap(cor_mat = res, col = my_colors)

```


```{r}

res.order <- res[order, order]

heatmap(res.order)

```

