---
title: "R Notebook"
output: html_notebook
---

```{r}

library(Seurat)
library(SeuratDisk)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(clustree)
library(reshape2)
library(tidyr)
library(tidyverse)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/xgboost_train.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/plottingFxns.R")

obj.opossum.glutamatergic <- readRDS("E:/Transcriptomics_M1/Opossum/seurat/opossum_m1_glutamatergic.rds")

```


```{r}

obj.opossum.glutamatergic <- ClusterSCT(obj.opossum.glutamatergic, c(1))
PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.1")

```


```{r}

canon.markers <- list(
                    class = c("Slc17a6", "Slc17a7", "Sv2b"),
                    L23 = c("Cux2", "Ccbe1", "Mdga1", "Stard8"),
                    L4 = c("Whrn", "Rorb"),
                    L5IT = c("Bcl11b"),
                    L5NP = c("Nxph1", "Tshz2", "Trhr", "Slc17a8"),
                    L5PT = c("Bcl6", "Erg", "Reln"),
                    L6CT = c("Foxp2", "Syt6"),
                    L6IT = c("Zfp804b", "Cdh9"),
                    L6b = c("Ctgf", "Inpp4b", "Svil")
                      )

# look for canonical markers
DefaultAssay(obj.opossum.glutamatergic) <- "RNA"
Idents(obj.opossum.glutamatergic) <- "SCT_snn_res.1"
PlotFeatures(obj.opossum.glutamatergic, canon.markers)
DotPlot(obj.opossum.glutamatergic, features = canon.markers, cols = c("lightgrey", "red"), scale = FALSE) + 
                         theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
                         theme(panel.background = element_rect(fill = "white", color = NA), plot.background = element_rect(fill = "white", color = NA)) + NoLegend()

```


```{r}

obj.opossum.glutamatergic.prefilt <- obj.opossum.glutamatergic
obj.opossum.glutamatergic <- subset(obj.opossum.glutamatergic, idents = c(NA), invert = TRUE)

```


```{r}

obj.opossum.glutamatergic <- ClusterSCT(obj.opossum.glutamatergic, c(1, 1.5, 2))
PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.1")

canon.markers <- list(
                    class = c("Slc17a6", "Slc17a7", "Sv2b"),
                    L23 = c("Cux2", "Ccbe1", "Mdga1", "Stard8"),
                    L4 = c("Whrn", "Rorb"),
                    L5IT = c("Bcl11b"),
                    L5NP = c("Nxph1", "Tshz2", "Trhr", "Slc17a8"),
                    L5PT = c("Bcl6", "Erg", "Reln"),
                    L6CT = c("Foxp2", "Syt6"),
                    L6IT = c("Zfp804b", "Cdh9"),
                    L6b = c("Ctgf", "Inpp4b", "Svil")
                      )

# look for canonical markers
DefaultAssay(obj.opossum.glutamatergic) <- "RNA"
Idents(obj.opossum.glutamatergic) <- "SCT_snn_res.1"
PlotFeatures(obj.opossum.glutamatergic, canon.markers)
DotPlot(obj.opossum.glutamatergic, features = canon.markers, cols = c("lightgrey", "red"), scale = FALSE) + 
                         theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
                         theme(panel.background = element_rect(fill = "white", color = NA), plot.background = element_rect(fill = "white", color = NA)) + NoLegend()

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.1.5")

```


```{r}

PlotClusters(obj.opossum.glutamatergic, group.id = "SCT_snn_res.2")

```


```{r}

clustree(obj.opossum.glutamatergic, prefix = "SCT_snn_res.")

```


```{r}

DefaultAssay(obj.opossum.glutamatergic) <- "SCT"
Idents(obj.opossum.glutamatergic) <- "SCT_snn_res.1"
markers.7 <- FindMarkers(obj.opossum.glutamatergic, ident.1 = 7)

```


```{r}

DefaultAssay(obj.opossum.glutamatergic) <- "SCT"
ident = "SCT_snn_res.1"
gene <- "Cck"
DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = ident, label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
FeaturePlot(obj.opossum.glutamatergic, gene, raster = FALSE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
VlnPlot(obj.opossum.glutamatergic, gene, group.by = ident)

```


```{r}

# identify subclasses at each clustering resolution
# do differential expression for clusters within each subclass
# generate plots of top genes for each

subclass.idx.opossum <- list()

subclass.idx.opossum$SCT_snn_res.1$IT <- c("2", "3", "4", "5", "7", "8", "9", "10", "11", "13", "14", "15", "17", "19")
subclass.idx.opossum$SCT_snn_res.1.5$IT <- c("1", "2", "3", "5", "6", "9", "10", "11", "12", "13", "14", "15", "16", "19", "22", "23", "24")
subclass.idx.opossum$SCT_snn_res.2$IT <- c("1", "3", "4", "6", "8", "9", "10", "11", "14", "15", "16", "17", "18", "19", "20", "21", "24", "25", "26")

subclass.idx.opossum$SCT_snn_res.1$L5PT <- c("6")
subclass.idx.opossum$SCT_snn_res.1.5$L5PT <- c("18", "21")
subclass.idx.opossum$SCT_snn_res.2$L5PT <- c("13", "23")

subclass.idx.opossum$SCT_snn_res.1$L5NP <- c("16")
subclass.idx.opossum$SCT_snn_res.1.5$L5NP <- c("17")
subclass.idx.opossum$SCT_snn_res.2$L5NP <- c("12")

subclass.idx.opossum$SCT_snn_res.1$L6CT <- c("1", "12")
subclass.idx.opossum$SCT_snn_res.1.5$L6CT <- c("4", "7", "8", "20")
subclass.idx.opossum$SCT_snn_res.2$L6CT <- c("2", "5", "7", "22")

subclass.idx.opossum$SCT_snn_res.1$L6b <- c("18")
subclass.idx.opossum$SCT_snn_res.1.5$L6b <- c("25")
subclass.idx.opossum$SCT_snn_res.2$L6b <- c("27")

obj.opossum.glutamatergic <- SubclassByIdent(obj.opossum.glutamatergic, subclass.idx.opossum)

```


```{r}

subclass.labels <- c("IT", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.1", "SCT_snn_res.1.5", "SCT_snn_res.2")

markers.opossum.glutamatergic <- IdentMarkerDict(obj.opossum.glutamatergic, subclass.labels, ident.labels,
                                        "E:/Transcriptomics_V1/Opossum/seurat/marker_genes/Glutamatergic/markerdict_clusters.rds")
SaveDotPlots(obj.opossum.glutamatergic, markers.opossum.glutamatergic,
             subclass.labels, ident.labels,
             "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/", "ENSMODG")
SaveFeaturePlots(obj.opossum.glutamatergic, markers.opossum.glutamatergic,
                 subclass.labels, ident.labels,
                 "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/")

```


```{r}

subclass.labels <- c("IT", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.1", "SCT_snn_res.1.5", "SCT_snn_res.2")

for (sbcl in subclass.labels) {
  for (id in ident.labels) {
    p <- PlotIdentGeneCounts(markers.opossum.glutamatergic, sbcl, id)
    p <- p + theme(aspect.ratio = 1) # Make plot square
    print(p)
  }
}

```


```{r}

DefaultAssay(obj.opossum.glutamatergic) <- "SCT"
ident = "SCT_snn_res.2"
gene <- "Pvalb"
DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = ident, label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
FeaturePlot(obj.opossum.glutamatergic, gene, raster = FALSE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
VlnPlot(obj.opossum.glutamatergic, gene, group.by = ident)

```


```{r}

# whole-transcriptome cross-validated classification
subclass.labels <- c("IT", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.1", "SCT_snn_res.1.5", "SCT_snn_res.2")

SaveIdentConfusionMatrices(obj.opossum.glutamatergic, subclass.labels, ident.labels, 
                      "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/")

```


```{r}

# label subclasses and types
obj.opossum.glutamatergic$subclass <- NA
obj.opossum.glutamatergic$type <- NA

subclass.resolutions <- rev(list(IT = 1, L5PT = 2, L5NP = 2, L6CT = 2, L6b = 2))
obj.opossum.glutamatergic <- LabelCells(obj.opossum.glutamatergic, subclass.resolutions)

```


```{r}

DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.opossum.glutamatergic, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

```


```{r}

# Plot relative subclass proportions
Idents(obj.opossum.glutamatergic) <- "subclass"
levels(obj.opossum.glutamatergic) <- c("IT", "L5PT", "L5NP", "L6CT", "L6b")

IdentBySample(obj.opossum.glutamatergic, y_limits = c(0, 0.60))

```


```{r}

subclass.labels <- c("IT", "L5PT", "L6CT")
ident.labels <- c("type")

subclass.markers.opossum.glutamatergic <- SubclassMarkerDict(obj.opossum.glutamatergic, "subclass",
                                                       "E:/Transcriptomics_V1/Opossum/seurat/marker_genes/Glutamatergic/markerdict_subclass.rds")
markers.opossum.glutamatergic <- IdentMarkerDict(obj.opossum.glutamatergic, subclass.labels, ident.labels,
                                        "E:/Transcriptomics_V1/Opossum/seurat/marker_genes/Glutamatergic/markerdict_types.rds")
SaveDotPlots(obj.opossum.glutamatergic, markers.opossum.glutamatergic,
             subclass.labels, ident.labels,
             "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/", "ENSMODG")
SaveFeaturePlots(obj.opossum.glutamatergic, markers.opossum.glutamatergic,
                 subclass.labels, ident.labels,
                 "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/")

PlotSubclassGeneCounts(subclass.markers.opossum.glutamatergic, "subclass", 
                       c("IT", "L5PT", "L5NP", "L6CT", "L6b")) + theme(aspect.ratio = 1)
for (sbcl in subclass.labels) {
  for (id in ident.labels) {
    p <- PlotIdentGeneCounts(markers.opossum.glutamatergic, sbcl, id)
    p <- p + theme(aspect.ratio = 1) # Make plot square
    print(p)
  }
}

SaveSubclassConfusionMatrices(obj.opossum.glutamatergic, "subclass",
                              c("IT", "L5PT", "L5NP", "L6CT", "L6b"),
                              "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/")
SaveIdentConfusionMatrices(obj.opossum.glutamatergic, subclass.labels, ident.labels,
                           "E:/Transcriptomics_V1/Opossum/seurat/cell_types/Glutamatergic/")

```


```{r}

saveRDS(obj.opossum.glutamatergic, "E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_glutamatergic_processed.rds")

```

