---
title: "R Notebook"
output: html_notebook
---


```{r}

library(Seurat)
library(SeuratDisk)
library(SeuratWrappers)
library(monocle3)
library(pheatmap)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(clustree)
library(reshape2)
library(tidyr)
library(gridExtra)
library(stringr)
library(plyr)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_integration_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/xgboost_train.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/plottingFxns.R")

obj.opossum.glutamatergic <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_glutamatergic_processed.rds")
obj.mouse.glutamatergic <- readRDS("E:/Transcriptomics_V1/Mouse/seurat/mouse_v1_P38_glutamatergic_processed.rds")

```


```{r}

colors_list <- list(
              # Glutamatergic
              IT = "#FF6C88",
              IT_A = "#FFB3B3",
              `L2/3` = "#FFB3B3",
              IT_B = "#FFA07A",
              L4 = "#FF7F50",
              IT_C = "#FF7F50",
              L5IT = "#FFA07A",
              IT_D = "#FF6347",
              L6IT = "#FF6347",
              L5NP = "#FF4500",
              L5PT = "#32a852", 
              L6CT = "#7de89a",
              L6b = "#206633",
            
              # GABAergic
              Pvalb = "#1E90FF",
              Sst = "#87CEEB",
              Vip = "#87CEFA",
              Lamp5 = "#4682B4",
              Frem1 = "#ADD8E6",
              Stac = "#5F9EA0",
            
              # Non-neuronal
              Astro = "#8C8C8C",
              Micro = "#A0A0A0",
              OD = "#B4B4B4",
              OPC = "#C8C8C8",
              Endo = "#505050",
              VLMC = "#B4B4B4"
)

```


```{r}

Idents(obj.opossum.glutamatergic) <- "subclass"
obj.opossum.IT <- subset(obj.opossum.glutamatergic, idents = c("IT_A", "IT_B", "IT_C", "IT_D"))

Idents(obj.mouse.glutamatergic) <- "subclass"
obj.mouse.IT <- subset(obj.mouse.glutamatergic, idents = c("L2/3", "L4", "L5IT", "L6IT"))

```


```{r}

obj.opossum.IT <- ClusterSCT(obj.opossum.IT, resolutions = c(1))
obj.mouse.IT <- ClusterSCT(obj.mouse.IT, resolutions = c(1))

```


```{r}

DimPlot(obj.opossum.IT, reduction = "umap", group.by = "sample", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.opossum.IT, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.opossum.IT, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
p <- DimPlot(obj.opossum.IT, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
ggsave("E:/Opossum_Paper/Figure 2/Opossum_V1_IT_UMAP_Subclass_FullSpace.png", plot = p, dpi = 300)
ggsave("E:/Opossum_Paper/Figure 2/Opossum_V1_IT_UMAP_Subclass_FullSpace.svg", plot = p, dpi = 300)

DimPlot(obj.mouse.IT, reduction = "umap", group.by = "sample", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.mouse.IT, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.mouse.IT, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
p <- DimPlot(obj.mouse.IT, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
ggsave("E:/Opossum_Paper/Figure 2/Mouse-P38_V1_IT_UMAP_Subclass_FullSpace.png", plot = p, dpi = 300)
ggsave("E:/Opossum_Paper/Figure 2/Mouse-P38_V1_IT_UMAP_Subclass_FullSpace.svg", plot = p, dpi = 300)

```


```{r}

opossum.vfs <- VariableFeatures(obj.opossum.IT)
mouse.vfs <- VariableFeatures(obj.mouse.IT)

poss.shared.vfs <- intersect(rownames(obj.opossum.IT), mouse.vfs)
shared.vfs <- intersect(opossum.vfs, mouse.vfs)
orthologues <- intersect(rownames(obj.opossum.IT), rownames(obj.mouse.IT))

```


```{r}

# cluster objects in shared VF space
obj.opossum.IT.SVF <- obj.opossum.IT
obj.mouse.IT.SVF <- obj.mouse.IT

obj.opossum.IT.SVF <- SCTransform(obj.opossum.IT.SVF, vst.flavor = "v2", residual.features = shared.vfs, return.only.var.genes = FALSE, verbose = FALSE) %>%
                      RunPCA(npcs = 30, verbose = FALSE) %>%
                      FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
                      RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE)

obj.mouse.IT.SVF <- SCTransform(obj.mouse.IT.SVF, vst.flavor = "v2", residual.features = shared.vfs, return.only.var.genes = FALSE, verbose = FALSE) %>%
                    RunPCA(npcs = 30, verbose = FALSE) %>%
                    FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
                    RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE)

```


```{r}

DimPlot(obj.opossum.IT.SVF, reduction = "umap", group.by = "sample", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.opossum.IT.SVF, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.opossum.IT.SVF, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
p <- DimPlot(obj.opossum.IT.SVF, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
ggsave("E:/Opossum_Paper/Figure 2/Opossum_V1_IT_UMAP_Subclass_SharedVFSpace.png", plot = p, dpi = 300)
ggsave("E:/Opossum_Paper/Figure 2/Opossum_V1_IT_UMAP_Subclass_SharedVFSpace.svg", plot = p, dpi = 300)

DimPlot(obj.mouse.IT.SVF, reduction = "umap", group.by = "sample", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.mouse.IT.SVF, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.mouse.IT.SVF, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
p <- DimPlot(obj.mouse.IT.SVF, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
ggsave("E:/Opossum_Paper/Figure 2/Mouse-P38_V1_IT_UMAP_Subclass_SharedVFSpace.png", plot = p, dpi = 300)
ggsave("E:/Opossum_Paper/Figure 2/Mouse-P38_V1_IT_UMAP_Subclass_SharedVFSpace.svg", plot = p, dpi = 300)

```


```{r}

# cluster objects in shared VF space
obj.opossum.IT.PVF <- obj.opossum.IT
obj.mouse.IT.PVF <- obj.mouse.IT

obj.opossum.IT.PVF <- SCTransform(obj.opossum.IT.SVF, vst.flavor = "v2", residual.features = poss.shared.vfs, return.only.var.genes = FALSE, verbose = FALSE) %>%
                      RunPCA(npcs = 30, verbose = FALSE) %>%
                      FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
                      RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE)

obj.mouse.IT.PVF <- SCTransform(obj.mouse.IT.SVF, vst.flavor = "v2", residual.features = poss.shared.vfs, return.only.var.genes = FALSE, verbose = FALSE) %>%
                    RunPCA(npcs = 30, verbose = FALSE) %>%
                    FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
                    RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE)

```


```{r}

DimPlot(obj.opossum.IT.PVF, reduction = "umap", group.by = "sample", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.opossum.IT.PVF, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.opossum.IT.PVF, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
p <- DimPlot(obj.opossum.IT.PVF, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
ggsave("E:/Opossum_Paper/Figure 2/Opossum_V1_IT_UMAP_Subclass_PossSharedVFSpace.png", plot = p, dpi = 300)
ggsave("E:/Opossum_Paper/Figure 2/Opossum_V1_IT_UMAP_Subclass_PossSharedVFSpace.svg", plot = p, dpi = 300)

DimPlot(obj.mouse.IT.PVF, reduction = "umap", group.by = "sample", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.mouse.IT.PVF, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.mouse.IT.PVF, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
p <- DimPlot(obj.mouse.IT.PVF, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE, shuffle = TRUE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
ggsave("E:/Opossum_Paper/Figure 2/Mouse-P38_V1_IT_UMAP_Subclass_PossSharedVFSpace.png", plot = p, dpi = 300)
ggsave("E:/Opossum_Paper/Figure 2/Mouse-P38_V1_IT_UMAP_Subclass_PossSharedVFSpace.svg", plot = p, dpi = 300)

```

