---
title: "R Notebook"
output: html_notebook
---

```{r}

library(Seurat)
library(SeuratDisk)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(clustree)
library(reshape2)
library(tidyr)
library(tidyverse)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/xgboost_train.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/plottingFxns.R")

obj.mouse.glutamatergic <- readRDS("E:/Transcriptomics_M1/Mouse/seurat/mouse_m1_glutamatergic.rds")

```


```{r}

obj.mouse.glutamatergic <- ClusterSCT(obj.mouse.glutamatergic, c(1))
obj.mouse.glutamatergic$predicted_doublets <- FALSE
obj.mouse.glutamatergic$predicted_doublets[1] <- TRUE
PlotClusters(obj.mouse.glutamatergic, group.id = "SCT_snn_res.1")

```


```{r}

canon.markers <- list(
                    class = c("Slc17a6", "Slc17a7", "Sv2b"),
                    L23 = c("Cux2", "Ccbe1", "Mdga1", "Stard8"),
                    L4 = c("Whrn", "Rorb"),
                    L5IT = c("Bcl11b"),
                    L5NP = c("Nxph1", "Tshz2", "Trhr", "Slc17a8"),
                    L5PT = c("Bcl6", "Erg", "Reln"),
                    L6CT = c("Foxp2", "Syt6"),
                    L6IT = c("Zfp804b", "Cdh9"),
                    L6b = c("Ctgf", "Inpp4b", "Svil")
                      )

# look for canonical markers
DefaultAssay(obj.mouse.glutamatergic) <- "RNA"
Idents(obj.mouse.glutamatergic) <- "SCT_snn_res.1"
PlotFeatures(obj.mouse.glutamatergic, canon.markers)
DotPlot(obj.mouse.glutamatergic, features = canon.markers, cols = c("lightgrey", "red"), scale = FALSE) + 
                         theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
                         theme(panel.background = element_rect(fill = "white", color = NA), plot.background = element_rect(fill = "white", color = NA)) + NoLegend()

```


```{r}

DimPlot(obj.mouse.glutamatergic, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.mouse.glutamatergic, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

```


```{r}

# identify subclasses at each clustering resolution
# do differential expression for clusters within each subclass
# generate plots of top genes for each

subclass.idx.mouse <- list()

subclass.idx.mouse$type$`L2/3` <- c("L2/3_A", "L2/3_B", "L2/3_C")

subclass.idx.mouse$type$L5IT <- c("L5IT_A", "L5IT_B", "L5IT_C", "L5IT_D", "L5IT_E", "L5IT_F")

subclass.idx.mouse$type$L6IT <- c("L6IT_A", "L6IT_B", "L6IT_C")

subclass.idx.mouse$type$L5PT <- c("L5PT_A", "L5PT_B", "L5PT_C", "L5PT_D")

subclass.idx.mouse$type$L5NP <- c("L5NP_A", "L5NP_B", "L5NP_C", "L5NP_D")

subclass.idx.mouse$type$L6CT <- c("L6CT_A", "L6CT_B", "L6CT_C", "L6CT_D")

subclass.idx.mouse$type$L6b <- c("L6b_A", "L6b_B", "L6b_C",  "L6b_D", "L6b_E")

obj.mouse.glutamatergic <- SubclassByIdent(obj.mouse.glutamatergic, subclass.idx.mouse)

```


```{r}

subclass.labels <- c("IT", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.1", "SCT_snn_res.1.5", "SCT_snn_res.2")

markers.mouse.glutamatergic <- IdentMarkerDict(obj.mouse.glutamatergic, subclass.labels, ident.labels,
                                        "E:/Transcriptomics_V1/Mouse/seurat/marker_genes/Glutamatergic/markerdict_clusters.rds")
SaveDotPlots(obj.mouse.glutamatergic, markers.mouse.glutamatergic,
             subclass.labels, ident.labels,
             "E:/Transcriptomics_V1/Mouse/seurat/cell_types/Glutamatergic/", "ENSMODG")
SaveFeaturePlots(obj.mouse.glutamatergic, markers.mouse.glutamatergic,
                 subclass.labels, ident.labels,
                 "E:/Transcriptomics_V1/Mouse/seurat/cell_types/Glutamatergic/")

```


```{r}

subclass.labels <- c("IT", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.1", "SCT_snn_res.1.5", "SCT_snn_res.2")

for (sbcl in subclass.labels) {
  for (id in ident.labels) {
    p <- PlotIdentGeneCounts(markers.mouse.glutamatergic, sbcl, id)
    p <- p + theme(aspect.ratio = 1) # Make plot square
    print(p)
  }
}

```


```{r}

DefaultAssay(obj.mouse.glutamatergic) <- "SCT"
ident = "SCT_snn_res.2"
gene <- "Pvalb"
DimPlot(obj.mouse.glutamatergic, reduction = "umap", group.by = ident, label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
FeaturePlot(obj.mouse.glutamatergic, gene, raster = FALSE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
VlnPlot(obj.mouse.glutamatergic, gene, group.by = ident)

```


```{r}

# whole-transcriptome cross-validated classification
subclass.labels <- c("IT", "L5PT", "L6CT")
ident.labels <- c("SCT_snn_res.1", "SCT_snn_res.1.5", "SCT_snn_res.2")

SaveIdentConfusionMatrices(obj.mouse.glutamatergic, subclass.labels, ident.labels, 
                      "E:/Transcriptomics_V1/Mouse/seurat/cell_types/Glutamatergic/")

```


```{r}

# label subclasses and types
obj.mouse.glutamatergic$subclass <- NA
obj.mouse.glutamatergic$type <- NA

subclass.resolutions <- rev(list(IT = 1, L5PT = 2, L5NP = 2, L6CT = 2, L6b = 2))
obj.mouse.glutamatergic <- LabelCells(obj.mouse.glutamatergic, subclass.resolutions)

```


```{r}

DimPlot(obj.mouse.glutamatergic, reduction = "umap", group.by = "subclass", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
DimPlot(obj.mouse.glutamatergic, reduction = "umap", group.by = "type", label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

```


```{r}

# Plot relative subclass proportions
Idents(obj.mouse.glutamatergic) <- "subclass"
levels(obj.mouse.glutamatergic) <- c("IT", "L5PT", "L5NP", "L6CT", "L6b")

IdentBySample(obj.mouse.glutamatergic, y_limits = c(0, 0.60))

```


```{r}

subclass.labels <- c("L2/3", "L5IT", "L5NP", "L5PT", "L6IT", "L6CT", "L6b")
ident.labels <- c("type")

subclass.markers.mouse.glutamatergic <- SubclassMarkerDict(obj.mouse.glutamatergic, "subclass",
                                                       "E:/Transcriptomics_M1/Mouse/seurat/marker_genes/Glutamatergic/markerdict_subclass.rds")
markers.mouse.glutamatergic <- IdentMarkerDict(obj.mouse.glutamatergic, subclass.labels, ident.labels,
                                        "E:/Transcriptomics_M1/Mouse/seurat/marker_genes/Glutamatergic/markerdict_types.rds")
SaveDotPlots(obj.mouse.glutamatergic, markers.mouse.glutamatergic,
             subclass.labels, ident.labels,
             "E:/Transcriptomics_M1/Mouse/seurat/cell_types/Glutamatergic/", "ENSMUSG")
SaveFeaturePlots(obj.mouse.glutamatergic, markers.mouse.glutamatergic,
                 subclass.labels, ident.labels,
                 "E:/Transcriptomics_M1/Mouse/seurat/cell_types/Glutamatergic/")

PlotSubclassGeneCounts(subclass.markers.mouse.glutamatergic, "subclass", 
                       c("L2/3", "L5IT", "L5NP", "L5PT", "L6IT", "L6CT", "L6b")) + theme(aspect.ratio = 1)
for (sbcl in subclass.labels) {
  for (id in ident.labels) {
    p <- PlotIdentGeneCounts(markers.mouse.glutamatergic, sbcl, id)
    p <- p + theme(aspect.ratio = 1) # Make plot square
    print(p)
  }
}

SaveSubclassConfusionMatrices(obj.mouse.glutamatergic, "subclass",
                              c("L2/3", "L5IT", "L5NP", "L5PT", "L6IT", "L6CT", "L6b"),
                              "E:/Transcriptomics_M1/Mouse/seurat/cell_types/Glutamatergic/")
SaveIdentConfusionMatrices(obj.mouse.glutamatergic, subclass.labels, ident.labels,
                           "E:/Transcriptomics_M1/Mouse/seurat/cell_types/Glutamatergic/")

```


```{r}

saveRDS(obj.mouse.glutamatergic, "E:/Transcriptomics_M1/Mouse/seurat/mouse_m1_glutamatergic_processed.rds")

```

